### PR TITLE
Warn on MIT-SHM not working on Linux X11

### PR DESCRIPTION
**Pull request by wesley800**
This PR simply throws a warn into log, when rustdesk server runs on an Linux box with its X11 server having MIT-SHM set to disabled. Plus another minor fix on pointer usage.
Personally I'm very unsatisfied about the current way of passing MIT-SHM errors from `scrap` to the server. Need some hint on API designing flavors.

### CODE REVIEW

**Comment by rustdesk** [comment]
This fixes bug or adds feature?

**Comment by wesley800** [comment]
> This fixes bug or adds feature?

I believe it's a feature, but only for debugging.

**Comment by rustdesk** [comment]
Debug for what?

**Comment by wesley800** [comment]
Currently when rustdesk runs server on SHM disabled environment, the client only sees black screen after connecting to the server, while keyboard and mouse works as normal. It is at least to me very confusing, when there's no relevant lines in rustdesk's log.
I tried to minimize the change. It also sounds reasonable to me to just throw a fatal error, cuz `scrap` requires SHM explicitly.

**Comment by rustdesk** [comment]
It is better to show here.

image

https://github.com/rustdesk/rustdesk/blob/286181ca04c43455ccf183badcf418cc9edbc231/flutter/lib/desktop/pages/desktop_home_page.dart#L324

**Comment by wesley800** [comment]
I agree to add a warn on GUI interface. What about headless setup? And I'm sorry that I can't write in flutter...

File: src/server/video_service.rs [hunk 1]
```diff
@@ -304,6 +309,16 @@ fn get_capturer(current: usize, portable_service_running: bool) -> ResultType<Ca
         );
     }
     let display = displays.remove(current);
+
+    if let Display::X11(inner) = &display {
+        if let Err(err) = inner.get_shm_status() {
```
Comment (rustdesk) [inline src/server/video_service.rs:315]: I do not think this can be compiled on non-Linux.

**Review by rustdesk — CHANGES_REQUESTED** [review]

**Comment by wesley800** [comment]
Could you explain a bit on why it's closed?

**Comment by rustdesk** [comment]
You did not answer https://github.com/rustdesk/rustdesk/pull/6856#discussion_r1452037313. I thought I had given up.

**Comment by wesley800** [comment]
I was expecting for ideas on how to pass the warns around, rather than a small `#[cfg]` modification. As I stated I have a strong feel that the PR needs refactor. Current commits are just for quick illustration and proof.

To be more specific, would you mind changing API of `scrap`? e.g. let `Capturer::new` returns `Result<(Capturer,WarningKind),ErrorKind>`, rather than `Result<Capturer,ErrorKind>`?

If API changing is not preferred, would you prefer the state (`shm_status` field) to be stored in `Server` instance, or checks on the fly each time `get_shm_status` is called?

**Comment by rustdesk** [comment]
I prefer `#[cfg]`

**Comment by wesley800** [comment]
BTW I'll try to add a tooltip to GUI, but it will take some time.

**Comment by rustdesk** [comment]
👍

**Comment by 109-999** [comment]
👍

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Pull request merged**